File: 1644-lowest-common-ancestor-of-a-binary-tree-ii/1644-lowest-common-ancestor-of-a-binary-tree-ii.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        //three pass
        TreeNode pNode = find(root, p);
        TreeNode qNode = find(root, q);
        if(pNode == null || qNode == null) return null;
        return LCA(root, p, q);
    }
    public TreeNode find(TreeNode root, TreeNode target) {
        if(root == null || target == null) return null;
        if(root == target) return root;
        TreeNode left = find(root.left, target);
        TreeNode right= find(root.right, target);
        return left == null ? right : left;
    }
    public TreeNode LCA(TreeNode root, TreeNode p, TreeNode q) {
        if(root == null || root == p || root == q) return root;
        TreeNode left = LCA(root.left, p, q);
        TreeNode right = LCA(root.right, p, q);
        if(left != null && right != null) return root;
        else if(left != null) return left;
        else if(right != null) return right;
        return null;
    }
}

File: 1644-lowest-common-ancestor-of-a-binary-tree-ii/1644-lowest-common-ancestor-of-a-binary-tree-ii.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        parent ={}
        parent[root] = None
        def dfs(root):
            nonlocal parent
            if not root:
                return 
            if root.left:
                parent[root.left] =root
                dfs(root.left)
            if root.right:
                parent[root.right] = root
                dfs(root.right)
        dfs(root)
        if not root or p not in parent or q not in parent:
            return None
        seen = set()
        while p:
            seen.add(p)
            p = parent[p]
        while q:
            if q in seen:
                return q
            q = parent[q]
        